### PR TITLE
Allow instance specification in`Cell::get_contained_cells`

### PR DESCRIPTION
**Pull request by pshriwise**
`Cell::get_contained_cells` returns the indices of all material-filled cells under a cell higher in the geometry hierarchy. Along with the indices is a vector of any instances of that cell as well. Currently, the instances returned are always representative of the first instance of the calling cell.

To make this method capable of locating contained cells and their instances for any instance of the higher-level cell (the calling cell), the calling cell's parent cells (the cells leading up to the root universe of the model) must be discovered as the full path from the root universe of the geometry to the cell of interest is required to correctly compute a cell's instance. This requires a non-trivial search for the correct path through the geometry to find this instance.

This PR implements that search to ensure that a full path from the top to the bottom of the geometry is completed when performing the calculation of contained cell instances. So in `get_contained_cells`, a `find_parent_cells` call generates a path _up_ the geometry levels and the `get_contained_cells_inner` generates all paths _down_ the geometry levels.

I've added a couple of convenience structures to make the search algorithm a little more digestible. One thing I'm not sure about  is whether or not the use of `openmc::vector` is appropriate throughout. Seeing as this is all internal, I opted to use it where possible but this might not make sense here.

To test this, I've added a redundant universe layer in the `cpp_driver` regression test to allow us to test this new capability by setting the cells in one section of the model's lattice to a specific temperature based on the specific instance of a cell in the newly inserted universe.

### CODE REVIEW

**Review by paulromano — CHANGES_REQUESTED** [review]
Tricky algorithm, but the implementation was pretty straightforward to follow. That search does seem like it could be costly depending on how the geometry is set up, but I'm not going to worry too much about that unless we actually find a case where it causes problems.

File: include/openmc/cell.h [hunk 1]
```diff
@@ -303,10 +310,62 @@ class UniversePartitioner {
...
 };
 
+struct ParentCellHash {
+  std::size_t operator()(const ParentCell& p) const
+  {
+    return 4096*p.cell_index + p.lattice_index;
+  }
+};
+
+struct ParentCellStack {
```
[the remark this answers is not in the document]
Comment (pshriwise) [reply include/openmc/cell.h:330]: Good call.

File: include/openmc/cell.h [hunk 2]
```diff
@@ -303,10 +310,62 @@ class UniversePartitioner {
...
+  { return cell_index == other.cell_index && lattice_index == other.lattice_index; }
+
+  bool operator<(const ParentCell& other) const
+  { return cell_index < other.cell_index || (cell_index == other.cell_index && lattice_index < other.lattice_index); }
+
   gsl::index cell_index;
   gsl::index lattice_index;
 };
 
+struct ParentCellHash {
```
[the remark this answers is not in the document]
Comment (pshriwise) [reply include/openmc/cell.h:323]: Thanks!
I noticed that these were missing for `CellInstance` and `CellInstanceHash` as well so I took the liberty of adding those too.

File: src/cell.cpp [hunk 3]
```diff
@@ -1220,10 +1220,106 @@ extern "C" int openmc_cell_set_name(int32_t index, const char* name)
   return 0;
 }
 
-std::unordered_map<int32_t, vector<int32_t>> Cell::get_contained_cells() const
+vector<ParentCell>
+Cell::find_parent_cells(vector<ParentCell>& parent_cells, int32_t instance) const
```
[the remark this answers is not in the document]
Comment (pshriwise) [reply src/cell.cpp:1224]: Oops! Yep. Changed the signature to something that made more sense right before submitting and forgot to remove this. Thanks!

File: include/openmc/cell.h [hunk 4]
```diff
@@ -166,10 +167,17 @@ class Cell {
   void set_name(const std::string& name) { name_ = name; };
 
   //! Get all cell instances contained by this cell
-  //! \return Map with cell indexes as keys and instances as values
-  std::unordered_map<int32_t, vector<int32_t>> get_contained_cells() const;
+  //! \param[in] instance Instance of the cell for which to get contained cells
+  //! (default instance is zero) \return Map with cell indexes as keys and
```
Comment (paulromano) [inline include/openmc/cell.h:171]: I think `\return` should be at the beginning of a new line

File: include/openmc/cell.h [hunk 5]
```diff
@@ -189,7 +197,7 @@ class Cell {
   //! \brief Index corresponding to this cell in distribcell arrays
   int distribcell_index_ {C_NONE};
 
-  //! \brief Material(s) within this cell.
+  //! \brief Material(s) within this cel
```
Comment (paulromano) [inline include/openmc/cell.h:200]:
```suggestion
  //! \brief Material(s) within this cell.
```

File: src/cell.cpp [hunk 6]
```diff
@@ -1220,10 +1220,185 @@ extern "C" int openmc_cell_set_name(int32_t index, const char* name)
...
+      // if we've updated the universe, break
+      if (prev_univ_idx != univ_idx) break;
+    } // end cell loop search for universe
+
+    // if we're at the top of the geometry and the instance matches, we're done
+    if (univ_idx == model::root_universe && stack.compute_instance(this->distribcell_index_) == instance) break;
+
+    // if there is no match on the original cell's universe, report an error
+    if (univ_idx == this->universe_) {
+      fatal_error(fmt::format("Could not find the parent cells for cell {}, instance {}.", this->id_, instance));
```
Comment (paulromano) [inline src/cell.cpp:1377]: Once again, looks like this file needs to be clang-formatted

Comment (pshriwise) [reply src/cell.cpp:1377]: Sorry about that. I've followed the instructions for installing the commit-hook that applies the style guide, but maybe I messed something up. 

```shell
All code changes were properly formatted
[find_parent_cells 1c1c848b3b4] Removing unused parameter in Cell::find_parent_cells.
 1 file changed, 27 insertions(+), 13 deletions(-)
```

I'll look into it.

**Pull request merged**